### ISSUE TITLE
[web] switch expression does not run synchronously on flutter test but on dart test

### ISSUE THREAD

**Issue by Tienisto** [flutter/flutter#138306]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

The switch statement runs synchronously on `dart test`, `dart test --platform chrome`, and `flutter test`.
However, it runs asynchronously on `flutter test --platform chrome`.

Please checkout the code snippet for more information.

### Expected results

`['Before switch', 'After switch', 'After run()']`

### Actual results

`['Before switch', 'After run()', 'After switch']`

### Code sample

<details><summary>Code sample</summary>

```dart
void main() {
  late FutureOr<int> value;
  setUp(() {
    value = 3;
  });

  test('Switch expression should run synchronously', () async {
    var order = <String>[];
    run() async {
      order.add('Before switch');
      final result = switch (value) {
        Future<int>() => await value,
        int() => value,
      };
      order.add('After switch');
      expect(result, 3);
    }

    var runFuture = run();
    order.add('After run()');

    await runFuture;
    expect(order, ['Before switch', 'After switch', 'After run()']);
  });
}
```

</details>

### Screenshots or Video

<details>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

<details><summary>Logs</summary>

```console
Expected: ['Before switch', 'After switch', 'After run()']
Actual: ['Before switch', 'After run()', 'After switch']
Which: at location [1] is 'After run()' instead of 'After switch'
```

</details>

### Flutter Doctor output

<details><summary>Doctor output</summary>

TLDR: I tested this on `3.13.6`, `3.13.9` and also on `3.16.0-0.5.pre`. All have the same issue.

```console
[!] Flutter (Channel beta, 3.16.0-0.5.pre, on macOS 13.5.2 22G91 darwin-arm64, locale en-GB)
    • Flutter version 3.16.0-0.5.pre on channel beta at /Users/tien/fvm/versions/3.16.0-0.5.pre
    ! Warning: `flutter` on your path resolves to /opt/homebrew/Caskroom/flutter/3.13.6/flutter/bin/flutter, which is not inside your current Flutter SDK checkout at /Users/tien/fvm/versions/3.16.0-0.5.pre. Consider adding /Users/tien/fvm/versions/3.16.0-0.5.pre/bin to the front of your path.
    ! Warning: `dart` on your path resolves to /opt/homebrew/Caskroom/flutter/3.13.6/flutter/bin/dart, which is not inside your current Flutter SDK checkout at /Users/tien/fvm/versions/3.16.0-0.5.pre. Consider adding /Users/tien/fvm/versions/3.16.0-0.5.pre/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision adc7dfe87e (4 days ago), 2023-11-07 18:30:33 -0800
    • Engine revision ab14aba1d5
    • Dart version 3.2.0 (build 3.2.0-210.4.beta)
    • DevTools version 2.28.1
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[!] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/tien/Library/Android/sdk
    ✗ cmdline-tools component is missing
      Run `path/to/sdkmanager --install "cmdline-tools;latest"`
      See https://developer.android.com/studio/command-line for more details.
    ✗ Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/macos#android-setup for more details.

[!] Xcode - develop for iOS and macOS (Xcode 14.3.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14E300c
    ✗ CocoaPods not installed.
        CocoaPods is used to retrieve the iOS and macOS platform side's plugin code that responds to your plugin usage on the Dart side.
        Without CocoaPods, plugins will not work on iOS or macOS.
        For more info, see https://flutter.dev/platform-plugins
      To install see https://guides.cocoapods.org/using/getting-started.html#installation for instructions.

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.3)
    • Android Studio at /Users/tien/Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)

[✓] IntelliJ IDEA Ultimate Edition (version 2023.2.1)
    • IntelliJ at /Users/tien/Applications/IntelliJ IDEA Ultimate.app
    • Flutter plugin version 75.1.4
    • Dart plugin version 232.9559.10

[✓] VS Code (version 1.81.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension can be installed from:
      🔨 https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-arm64   • macOS 13.5.2 22G91 darwin-arm64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 119.0.6045.123

[✓] Network resources
    • All expected network resources are available.
```

</details>

**Comment by huycozy** [comment]
Hi @Tienisto 
I see it can be solved if you `await` for `run()` before adding `After run()`:

```dart
var runFuture = await run();
order.add('After run()');
```

Please try to see if it works for your case and confirm this.

**Comment by Tienisto** [comment]
Hello, the main issue is that `after run` should be added last.

This is due to the fact that the switch expression shall not add another async task because the "await" branch is not executed.

**Comment by huycozy** [comment]
Reproduced the issue in master channel as well. 

<details>
<summary>flutter doctor -v (master)</summary>

```bash
[!] Flutter (Channel master, 3.17.0-5.0.pre.80, on macOS 14.1 23B74 darwin-x64, locale en-VN)
    • Flutter version 3.17.0-5.0.pre.80 on channel master at /Users/huynq/Documents/GitHub/flutter_master
    ! Warning: `flutter` on your path resolves to /Users/huynq/Documents/GitHub/flutter/bin/flutter, which is not inside your current Flutter SDK checkout at /Users/huynq/Documents/GitHub/flutter_master. Consider adding /Users/huynq/Documents/GitHub/flutter_master/bin to the front of your path.
    ! Warning: `dart` on your path resolves to /Users/huynq/Documents/GitHub/flutter/bin/dart, which is not inside your current Flutter SDK checkout at /Users/huynq/Documents/GitHub/flutter_master. Consider adding /Users/huynq/Documents/GitHub/flutter_master/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 6e9d4ccc25 (21 minutes ago), 2023-11-13 21:46:14 -0500
    • Engine revision d100e8912e
    • Dart version 3.3.0 (build 3.3.0-127.0.dev)
    • DevTools version 2.30.0-dev.4
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/huynq/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • ANDROID_HOME = /Users/huynq/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15A507
    • CocoaPods version 1.13.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.3)
    • Android Studio at /Applications/Android Studio Giraffe.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)

[!] Android Studio (version unknown)
    • Android Studio at /Applications/Android Studio Preview.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    ✗ Unable to determine Android Studio version.
    • Java version OpenJDK Runtime Environment (build 17.0.8+0-17.0.8b1000.22-10799086)

[✓] Android Studio (version 2022.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • android-studio-dir = /Applications/Android Studio.app/
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)

[✓] VS Code (version 1.84.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.76.0

[✓] Connected device (3 available)
    • RMX2001 (mobile) • EUYTFEUSQSRGDA6D • android-arm64  • Android 11 (API 30)
    • macOS (desktop)  • macos            • darwin-x64     • macOS 14.1 23B74 darwin-x64
    • Chrome (web)     • chrome           • web-javascript • Google Chrome 119.0.6045.123

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 2 categories.
```
</details>

**Comment by mraleph** [comment]
dart2js vs DDC descrepancy perhaps? /cc @sigmundch @nshahan

**Comment by nshahan** [comment]
@mraleph Yeah it looks like DDC has a discrepancy here. I opened https://github.com/dart-lang/sdk/issues/54050 to track.

@Tienisto Thanks for the reproduction! Could you help us determine the priority of this bug by describing where you discovered it and if you have a workaround? Depending on how close your code is to the repro you could possibly avoid this ordering issues with a ternary expression instead of the switch case: 
```dart
final result = value is int ? value : await value;
```

**Comment by Tienisto** [comment]
I discovered it because I called a method in `initState` of a `State` expecting that it is finished before calling `build`. So it is just an edge case.

**Comment by yjbanov** [comment]
Closing in favor of https://github.com/dart-lang/sdk/issues/54050
